File: tools/verifier-ts/src/parse.ts
// Strict canonical NDJSON-line parser for ODC events.
//
// Built from contracts/ alone: event-schema.md (ES-*), export-format.md (EX-*),
// hashing.md (HA-2, HA-6), event-types.md (ET-14 uses the decoded scalars this
// produces). This parser enforces the *single valid byte representation* an
// event line MUST have (EX-7..EX-10) directly on the raw bytes — it does NOT
// delegate to JSON.parse, which silently accepts duplicate keys, non-canonical
// numbers (1e2, 1.0), and loses key order, all of which the spec rejects (D5).
//
// A successful parse means every single-line Stage A check has passed:
//   - compact object, exactly the 7 envelope fields in fixed order (EX-7, ES-1/2)
//   - no null / absent field (ES-3)
//   - seq/version canonical integers in [0, 2^53-1] (ES-5, ES-12)
//   - type matches ^[a-z][a-z0-9_]*$ (ES-10)
//   - payload flat, values integer-or-string only (ES-15/16/17, EV-16),
//     keys strictly ascending by UTF-8 bytes, no duplicates (EX-8, HA-6)
//   - every string minimally + canonically escaped, valid UTF-8 (EX-9, HA-2)
//   - ts syntactically + calendrically valid (ES-20)
//   - prev_hash / hash are 64 lowercase hex (ES-23, ES-26)
//
// Any violation throws ParseFail; parseEventLine catches it and returns null,
// which the verifier renders as INVALID at that line. Cross-line checks
// (seq contiguity, prev_hash linkage, genesis position, hash recomputation) and
// all type semantics live in verify.ts.

export type PayloadValue =
  { kind: "int"; value: number } | { kind: "str"; value: string };

export interface PayloadEntry {
  key: string; // decoded scalar value of the key
  keyBytes: Buffer; // UTF-8 octets of the decoded key (HA-8 ordering)
  val: PayloadValue;
}

export interface ParsedEvent {
  seq: number;
  type: string;
  version: number;
  payload: PayloadEntry[]; // in stored (ascending) order
  ts: string;
  prevHash: string;
  hash: string;
}

const MAX_INT = 9007199254740991; // 2^53 - 1 (ES-5)

class ParseFail extends Error {}

// Reader over the raw line bytes. Never skips whitespace: the canonical form is
// compact (EX-7), so any space/tab/newline between tokens is a violation.
class Reader {
  constructor(
    private readonly b: Buffer,
    public pos = 0,
  ) {}

  atEnd(): boolean {
    return this.pos >= this.b.length;
  }

  peek(): number {
    if (this.pos >= this.b.length) throw new ParseFail("eof");
    // Buffer indexing is a number here; noUncheckedIndexedAccess widens it.
    return this.b[this.pos] as number;
  }

  next(): number {
    const c = this.peek();
    this.pos++;
    return c;
  }

  expect(byte: number): void {
    if (this.next() !== byte) throw new ParseFail("expected byte");
  }

  // The number token is pure ASCII (digits/sign/exponent), so latin1 is exact.
  slice(from: number, to: number): string {
    return this.b.subarray(from, to).toString("latin1");
  }
}

const isDigit = (c: number): boolean => c >= 0x30 && c <= 0x39;
const isLowerHex = (c: number): boolean =>
  (c >= 0x30 && c <= 0x39) || (c >= 0x61 && c <= 0x66);

const HEX64 = /^[0-9a-f]{64}$/;
const TYPE_RE = /^[a-z][a-z0-9_]*$/;

// --- string parsing (EX-9 canonical escaping, HA-2 UTF-8) --------------------

function parseString(r: Reader): { value: string; bytes: Buffer } {
  r.expect(0x22); // opening quote
  const cps: number[] = [];
  for (;;) {
    const c = r.next();
    if (c === 0x22) break; // closing quote
    if (c === 0x5c) {
      // backslash escape
      const e = r.next();
      switch (e) {
        case 0x22: // \"
          cps.push(0x22);
          break;
        case 0x5c: // \\
          cps.push(0x5c);
          break;
        case 0x62: // \b
          cps.push(0x08);
          break;
        case 0x74: // \t
          cps.push(0x09);
          break;
        case 0x6e: // \n
          cps.push(0x0a);
          break;
        case 0x66: // \f
          cps.push(0x0c);
          break;
        case 0x72: // \r
          cps.push(0x0d);
          break;
        case 0x75: {
          // \u00xx — ONLY the C0 controls that lack a short escape, lowercase hex
          const h0 = r.next();
          const h1 = r.next();
          const h2 = r.next();
          const h3 = r.next();
          if (
            !isLowerHex(h0) ||
            !isLowerHex(h1) ||
            !isLowerHex(h2) ||
            !isLowerHex(h3)
          ) {
            throw new ParseFail("non-lowercase-hex \\u escape");
          }
          const v = parseInt(String.fromCharCode(h0, h1, h2, h3), 16);
          // Minimal escaping: \u is legal ONLY for U+0000..U+001F excluding the
          // five that MUST use a short escape (\b \t \n \f \r). Everything else
          // — including non-controls and U+007F+ — is literal, never \u (EX-9).
          const shortEscaped =
            v === 0x08 || v === 0x09 || v === 0x0a || v === 0x0c || v === 0x0d;
          if (v > 0x1f || shortEscaped) {
            throw new ParseFail("non-minimal \\u escape");
          }
          cps.push(v);
          break;
        }
        default:
          // Any other escape (\/ among them) is non-canonical (EX-9).
          throw new ParseFail("illegal escape");
      }
      continue;
    }
    if (c < 0x20) {
      // A raw control byte MUST have been escaped (EX-9).
      throw new ParseFail("raw control byte in string");
    }
    if (c < 0x80) {
      cps.push(c); // ASCII, incl. 0x7f which EX-9 leaves literal
      continue;
    }
    // Multi-byte UTF-8, decoded and validated (HA-2 rejects ill-formed UTF-8).
    let need: number;
    let cp: number;
    let min: number;
    if (c >= 0xc2 && c <= 0xdf) {
      need = 1;
      cp = c & 0x1f;
      min = 0x80;
    } else if (c >= 0xe0 && c <= 0xef) {
      need = 2;
      cp = c & 0x0f;
      min = 0x800;
    } else if (c >= 0xf0 && c <= 0xf4) {
      need = 3;
      cp = c & 0x07;
      min = 0x10000;
    } else {
      // 0x80..0xC1 (continuation or overlong lead) and 0xF5..0xFF are invalid.
      throw new ParseFail("invalid UTF-8 lead byte");
    }
    for (let i = 0; i < need; i++) {
      const cont = r.next();
      if (cont < 0x80 || cont > 0xbf) throw new ParseFail("bad continuation");
      cp = (cp << 6) | (cont & 0x3f);
    }
    if (cp < min) throw new ParseFail("overlong UTF-8");
    if (cp >= 0xd800 && cp <= 0xdfff) throw new ParseFail("surrogate in UTF-8");
    if (cp > 0x10ffff) throw new ParseFail("UTF-8 out of range");
    cps.push(cp);
  }
  const value = String.fromCodePoint(...cps);
  return { value, bytes: Buffer.from(value, "utf8") };
}

// --- number parsing (ES-5 canonical integer) ---------------------------------

const CANON_INT = /^(0|[1-9][0-9]*)$/;

function parseCanonicalInteger(r: Reader): number {
  const start = r.pos;
  // Collect the full JSON number token so a float/exponent/sign is recognised
  // and then rejected, rather than silently truncated at the '.'.
  while (!r.atEnd()) {
    const c = r.peek();
    if (
      isDigit(c) ||
      c === 0x2d || // -
      c === 0x2b || // +
      c === 0x2e || // .
      c === 0x65 || // e
      c === 0x45 // E
    ) {
      r.pos++;
    } else {
      break;
    }
  }
  const raw = r.slice(start, r.pos);
  if (!CANON_INT.test(raw)) throw new ParseFail("non-canonical integer");
  const big = BigInt(raw);
  if (big > BigInt(MAX_INT)) throw new ParseFail("integer out of range");
  return Number(big);
}

// --- ts calendar validation (ES-20) ------------------------------------------

const TS_RE = /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})\.(\d{3})Z$/;

function isLeap(y: number): boolean {
  return (y % 4 === 0 && y % 100 !== 0) || y % 400 === 0;
}

function validTimestamp(s: string): boolean {
  const m = TS_RE.exec(s);
  if (!m) return false; // syntactic gate
  const year = Number(m[1]);
  const month = Number(m[2]);
  const day = Number(m[3]);
  const hour = Number(m[4]);
  const min = Number(m[5]);
  const sec = Number(m[6]);
  // Calendar gate: real UTC instant; leap seconds (60) rejected (ES-20).
  if (month < 1 || month > 12) return false;
  const dim = [
    31,
    isLeap(year) ? 29 : 28,
    31,
    30,
    31,
    30,
    31,
    31,
    30,
    31,
    30,
    31,
  ];
  const maxDay = dim[month - 1] as number;
  if (day < 1 || day > maxDay) return false;
  if (hour > 23) return false;
  if (min > 59) return false;
  if (sec > 59) return false;
  return true;
}

// --- payload parsing (EX-8 ordering, HA-6 duplicates, ES-16/17 value types) ---

function parsePayload(r: Reader): PayloadEntry[] {
  r.expect(0x7b); // {
  const entries: PayloadEntry[] = [];
  if (r.peek() === 0x7d) {
    r.pos++; // empty payload {}
    return entries;
  }
  let prevKeyBytes: Buffer | null = null;
  for (;;) {
    const key = parseString(r);
    r.expect(0x3a); // :
    // Ordering + duplicate check on the raw decoded key bytes (HA-8 / EX-8).
    if (prevKeyBytes !== null) {
      const cmp = Buffer.compare(prevKeyBytes, key.bytes);
      if (cmp === 0) throw new ParseFail("duplicate payload key"); // HA-6
      if (cmp > 0) throw new ParseFail("payload keys not ascending"); // EX-8
    }
    prevKeyBytes = key.bytes;

    const c = r.peek();
    let val: PayloadValue;
    if (c === 0x22) {
      val = { kind: "str", value: parseString(r).value };
    } else if (isDigit(c) || c === 0x2d) {
      // A leading '-' is captured so negative / non-canonical forms are rejected
      // by parseCanonicalInteger rather than mis-parsed.
      val = { kind: "int", value: parseCanonicalInteger(r) };
    } else {
      // { [ t f n  → nested object/array, boolean, or null: all illegal in a
      // v1 payload (ES-16/ES-17). EV-16 makes this INVALID even on an
      // unregistered type, because such a value has no HA-7 encoding.
      throw new ParseFail("illegal payload value type");
    }
    entries.push({ key: key.value, keyBytes: key.bytes, val });

    const sep = r.next();
    if (sep === 0x7d) break; // }
    if (sep !== 0x2c) throw new ParseFail("expected , or }"); // ,
  }
  return entries;
}

// --- top-level event line ----------------------------------------------------

const ENVELOPE_ORDER = [
  "seq",
  "type",
  "version",
  "payload",
  "ts",
  "prev_hash",
  "hash",
] as const;

function parseEnvelopeKey(r: Reader, expected: string): void {
  const k = parseString(r);
  if (k.value !== expected) throw new ParseFail("wrong envelope key/order");
  r.expect(0x3a); // :
}

function parseEvent(r: Reader): ParsedEvent {
  r.expect(0x7b); // {

  // seq
  parseEnvelopeKey(r, ENVELOPE_ORDER[0]);
  const seq = parseCanonicalInteger(r);
  r.expect(0x2c);

  // type
  parseEnvelopeKey(r, ENVELOPE_ORDER[1]);
  const type = parseString(r).value;
  if (!TYPE_RE.test(type)) throw new ParseFail("malformed type (ES-10)");
  r.expect(0x2c);

  // version
  parseEnvelopeKey(r, ENVELOPE_ORDER[2]);
  const version = parseCanonicalInteger(r);
  if (version < 1) throw new ParseFail("version < 1 (ES-12)");
  r.expect(0x2c);

  // payload
  parseEnvelopeKey(r, ENVELOPE_ORDER[3]);
  const payload = parsePayload(r);
  r.expect(0x2c);

  // ts
  parseEnvelopeKey(r, ENVELOPE_ORDER[4]);
  const ts = parseString(r).value;
  if (!validTimestamp(ts)) throw new ParseFail("bad ts (ES-20)");
  r.expect(0x2c);

  // prev_hash
  parseEnvelopeKey(r, ENVELOPE_ORDER[5]);
  const prevHash = parseString(r).value;
  if (!HEX64.test(prevHash)) throw new ParseFail("bad prev_hash (ES-23)");
  r.expect(0x2c);

  // hash
  parseEnvelopeKey(r, ENVELOPE_ORDER[6]);
  const hash = parseString(r).value;
  if (!HEX64.test(hash)) throw new ParseFail("bad hash (ES-26)");

  r.expect(0x7d); // }
  if (!r.atEnd()) throw new ParseFail("trailing bytes after object");

  return { seq, type, version, payload, ts, prevHash, hash };
}

/**
 * Parse one raw NDJSON line (without its terminating LF) into a ParsedEvent, or
 * null if it violates any single-line Stage A / canonical-form rule. Returning
 * null (not throwing) lets the caller attribute INVALID to this line number.
 */
export function parseEventLine(lineBytes: Buffer): ParsedEvent | null {
  try {
    return parseEvent(new Reader(lineBytes));
  } catch (e) {
    if (e instanceof ParseFail) return null;
    // A non-ParseFail (e.g. String.fromCodePoint range) still means the line is
    // not a conforming event; treat as INVALID rather than crashing the tool.
    return null;
  }
}

File: services/verifier/internal/verify/parse.go
package verify

import "unicode/utf8"

// This file implements a strict, byte-exact JSON parser used to enforce the
// canonical line form of contracts/export-format.md (EX-7..EX-10) while
// extracting field values. It deliberately does NOT use encoding/json: the
// standard decoder tolerates whitespace, reorders keys, silently accepts
// duplicate keys, and normalizes numbers — every one of which the canonical
// form forbids. The parser rejects any deviation from the compact canonical
// serialization by construction (no whitespace is ever skipped).

type jkind int

const (
	kObject jkind = iota
	kString
	kInt       // canonical non-negative integer in [0, 2^53-1] (ES-5)
	kBadNumber // any numeric token that is NOT a canonical in-range integer
	kBool
	kNull
	kArray
)

type jvalue struct {
	kind jkind
	str  string // decoded string value (kString)
	ival int64  // integer value (kInt)
	obj  *jobject
}

type jobject struct {
	keys   []string // decoded keys, in stored order
	vals   []jvalue
	dupKey bool // set if any key appears more than once (HA-6)
}

type parser struct {
	b   []byte
	pos int
	err bool
}

// parseObjectLine parses exactly one JSON object occupying the whole input,
// with no leading, trailing, or interior whitespace. It returns ok=false for
// any structural or canonical-form violation.
func parseObjectLine(b []byte) (*jobject, bool) {
	if len(b) == 0 || b[0] != '{' {
		return nil, false
	}
	p := &parser{b: b}
	v := p.parseValue()
	if p.err || p.pos != len(b) || v.kind != kObject {
		return nil, false
	}
	return v.obj, true
}

func (p *parser) parseValue() jvalue {
	if p.err || p.pos >= len(p.b) {
		p.err = true
		return jvalue{}
	}
	c := p.b[p.pos]
	switch {
	case c == '{':
		return p.parseObject()
	case c == '[':
		return p.parseArray()
	case c == '"':
		s, ok := p.parseString()
		if !ok {
			p.err = true
			return jvalue{}
		}
		return jvalue{kind: kString, str: s}
	case c == 't':
		return p.parseLit("true", kBool)
	case c == 'f':
		return p.parseLit("false", kBool)
	case c == 'n':
		return p.parseLit("null", kNull)
	case c == '-' || (c >= '0' && c <= '9'):
		return p.parseNumber()
	default:
		p.err = true
		return jvalue{}
	}
}

// dupSetThreshold is the key count at which duplicate detection switches from
// a linear scan of the keys already seen to a set. Below it the scan is
// cheaper than allocating a map, and real ODC events are far below it — the
// widest v1 payload has five keys. Above it the scan is what made a wide
// payload quadratic in key count, so a hostile export of very modest size
// could wedge the verifier: 200k keys took over a minute of pure key
// comparison. "A stranger can write a verifier and check the log in an
// afternoon" (charter §4) does not survive that, so the set takes over.
const dupSetThreshold = 32

func (p *parser) parseObject() jvalue {
	p.pos++ // consume '{'
	obj := &jobject{}
	if p.pos < len(p.b) && p.b[p.pos] == '}' {
		p.pos++
		return jvalue{kind: kObject, obj: obj}
	}
	// seen is nil until the key count crosses dupSetThreshold, at which point
	// it is populated from the keys already stored and used from then on.
	var seen map[string]struct{}
	for {
		if p.pos >= len(p.b) || p.b[p.pos] != '"' {
			p.err = true
			return jvalue{}
		}
		key, ok := p.parseString()
		if !ok {
			p.err = true
			return jvalue{}
		}
		if p.pos >= len(p.b) || p.b[p.pos] != ':' {
			p.err = true
			return jvalue{}
		}
		p.pos++ // consume ':'
		val := p.parseValue()
		if p.err {
			return jvalue{}
		}
		// HA-6 duplicate-key detection. Once dupKey is set the answer cannot
		// change, so we stop looking entirely: dupKey is a single boolean the
		// caller reads, not a list of offending keys, and continuing to search
		// after the first hit buys nothing. This is both the early exit the
		// previous scan lacked and, with the set below, what makes the whole
		// thing sub-quadratic. Parsing itself continues either way, because a
		// duplicate key is not a parse error — it is a Stage A rejection the
		// caller makes after the line is fully parsed, and stopping here would
		// change which check fires (though not, per EV-17, which line).
		if !obj.dupKey {
			switch {
			case seen != nil:
				if _, dup := seen[key]; dup {
					obj.dupKey = true
				} else {
					seen[key] = struct{}{}
				}
			default:
				for _, k := range obj.keys {
					if k == key {
						obj.dupKey = true
						break
					}
				}
				if !obj.dupKey && len(obj.keys)+1 >= dupSetThreshold {
					seen = make(map[string]struct{}, len(obj.keys)+1)
					for _, k := range obj.keys {
						seen[k] = struct{}{}
					}
					seen[key] = struct{}{}
				}
			}
		}
		obj.keys = append(obj.keys, key)
		obj.vals = append(obj.vals, val)
		if p.pos >= len(p.b) {
			p.err = true
			return jvalue{}
		}
		switch p.b[p.pos] {
		case ',':
			p.pos++
			continue
		case '}':
			p.pos++
			return jvalue{kind: kObject, obj: obj}
		default:
			p.err = true
			return jvalue{}
		}
	}
}

// parseArray fully consumes a JSON array so parsing can continue; the array
// itself is never valid in a v1 payload (ES-17) but must be recognized.
func (p *parser) parseArray() jvalue {
	p.pos++ // consume '['
	if p.pos < len(p.b) && p.b[p.pos] == ']' {
		p.pos++
		return jvalue{kind: kArray}
	}
	for {
		p.parseValue()
		if p.err {
			return jvalue{}
		}
		if p.pos >= len(p.b) {
			p.err = true
			return jvalue{}
		}
		switch p.b[p.pos] {
		case ',':
			p.pos++
			continue
		case ']':
			p.pos++
			return jvalue{kind: kArray}
		default:
			p.err = true
			return jvalue{}
		}
	}
}

func (p *parser) parseLit(lit string, k jkind) jvalue {
	if p.pos+len(lit) <= len(p.b) && string(p.b[p.pos:p.pos+len(lit)]) == lit {
		p.pos += len(lit)
		return jvalue{kind: k}
	}
	p.err = true
	return jvalue{}
}

func (p *parser) parseNumber() jvalue {
	start := p.pos
	for p.pos < len(p.b) {
		c := p.b[p.pos]
		if (c >= '0' && c <= '9') || c == '-' || c == '+' || c == '.' || c == 'e' || c == 'E' {
			p.pos++
		} else {
			break
		}
	}
	tok := p.b[start:p.pos]
	if v, ok := canonicalInt(tok); ok {
		return jvalue{kind: kInt, ival: v}
	}
	return jvalue{kind: kBadNumber}
}

// maxSafeInteger is 2^53 - 1 (ES-5): integers must round-trip losslessly.
const maxSafeInteger int64 = 9007199254740991

// canonicalInt reports whether tok is the canonical integer form of ES-5
// (^(0|[1-9][0-9]*)$) and lies within [0, 2^53-1], returning its value.
func canonicalInt(tok []byte) (int64, bool) {
	if len(tok) == 0 {
		return 0, false
	}
	if tok[0] == '0' {
		if len(tok) != 1 { // leading zero
			return 0, false
		}
		return 0, true
	}
	var v int64
	for _, c := range tok {
		if c < '0' || c > '9' {
			return 0, false
		}
		v = v*10 + int64(c-'0')
		if v > maxSafeInteger {
			return 0, false
		}
	}
	return v, true
}

// parseString parses a JSON string starting at a '"', enforcing EX-9 minimal
// escaping and HA-2 well-formed-UTF-8, and returns the decoded scalar-value
// string (the value UTF8() hashes over).
func (p *parser) parseString() (string, bool) {
	if p.pos >= len(p.b) || p.b[p.pos] != '"' {
		return "", false
	}
	p.pos++ // consume opening quote
	var out []byte
	for {
		if p.pos >= len(p.b) {
			return "", false // unterminated
		}
		c := p.b[p.pos]
		switch {
		case c == '"':
			p.pos++
			if !utf8.Valid(out) {
				return "", false
			}
			return string(out), true
		case c == '\\':
			if p.pos+1 >= len(p.b) {
				return "", false
			}
			e := p.b[p.pos+1]
			switch e {
			case '"':
				out = append(out, '"')
				p.pos += 2
			case '\\':
				out = append(out, '\\')
				p.pos += 2
			case 'b':
				out = append(out, 0x08)
				p.pos += 2
			case 't':
				out = append(out, 0x09)
				p.pos += 2
			case 'n':
				out = append(out, 0x0a)
				p.pos += 2
			case 'f':
				out = append(out, 0x0c)
				p.pos += 2
			case 'r':
				out = append(out, 0x0d)
				p.pos += 2
			case 'u':
				if p.pos+6 > len(p.b) {
					return "", false
				}
				v, ok := lowerHex4(p.b[p.pos+2 : p.pos+6])
				if !ok {
					return "", false // non-hex or uppercase (EX-9)
				}
				// \u is permitted ONLY for control chars U+0000-U+001F that are
				// not one of the five with mandatory short escapes.
				if v >= 0x20 {
					return "", false // non-ASCII / >=0x20 must be literal (EX-9)
				}
				if v == 0x08 || v == 0x09 || v == 0x0a || v == 0x0c || v == 0x0d {
					return "", false // must use short escape (non-minimal)
				}
				out = append(out, byte(v))
				p.pos += 6
			default:
				return "", false // invalid escape (incl. \/)
			}
		case c < 0x20:
			return "", false // raw control must be escaped
		default:
			out = append(out, c) // literal byte: ASCII >=0x20, 0x7f, or UTF-8 lead/cont
			p.pos++
		}
	}
}

// lowerHex4 decodes exactly four lowercase-hex digits.
func lowerHex4(h []byte) (int, bool) {
	if len(h) != 4 {
		return 0, false
	}
	v := 0
	for _, c := range h {
		var d int
		switch {
		case c >= '0' && c <= '9':
			d = int(c - '0')
		case c >= 'a' && c <= 'f':
			d = int(c-'a') + 10
		default:
			return 0, false // uppercase or non-hex
		}
		v = v*16 + d
	}
	return v, true
}
